stablehlo/tests/CheckOps.cpp

```diff
@@ -156,6 +157,96 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensor);
 }
 
+static uint64_t ULPDifference(APFloat f, APFloat g) {
+  // z is ULP-distance between exact 0 and largest subnormal:
+  uint64_t z = APFloat::getSmallestNormalized(f.getSemantics())
+                   .bitcastToAPInt()
+                   .getLimitedValue() -
+               1;
+  // a is ULP-distance between exact 0 and abs(f):
+  uint64_t a = (f.isNegative() ? -f : f).bitcastToAPInt().getLimitedValue();
+  // b is ULP-distance between exact 0 and abs(g):
+  uint64_t b = (g.isNegative() ? -g : g).bitcastToAPInt().getLimitedValue();
+  if (f.isFinite() && g.isFinite()) {
+    // subtract subnormals contribution, round subnormals to closest normal or
+    // zero:
+    if (z < a)
+      a -= z;
+    else
+      a = (2 * a < z ? 0 : 1);
+    if (z < b)
+      b -= z;
+    else
+      b = (2 * b < z ? 0 : 1);
+    if (f.isNegative() != g.isNegative()) {
+      // in case of overflow, return uint64 maximal value
+      if (a > std::numeric_limits<uint64_t>::max() - b)
+        return std::numeric_limits<uint64_t>::max();
```

Comment:
This can never overflow, given that the most significant bit of `a` and `b` are both zero, since the sign bit is zero.

Reply:
Nice catch! Indeed, when inputs are finite, then `a` and `b` are the uint views of the absolute values of inputs that will not exceed `std::numeric_limits<uint64_t>::max()` in the cases of `float` and `double` inputs. However, for 'long double' inputs, `a + b` could still overflow.. Resolving with the assert suggestion in https://github.com/openxla/stablehlo/pull/2357#discussion_r1635403517 and deleting this if-block.

Reply:
Fixed.